zcash_client_backend/src/data_api/chain.rs

```diff
@@ -134,59 +134,52 @@ pub trait BlockSource {
 /// block source is more likely to be accurate than the previously-scanned information.
 /// This follows from the design (and trust) assumption that the `lightwalletd` server
 /// provides accurate block information as of the time it was requested.
+/// The addition of a `limit` to the chain validation function changes the meaning of
+/// its successful output, being now a `BlockHeight, BlockHash)` tuple indicating the
+/// block height and block hash up to which the chain as been validated on its continuity
+/// of heights and hashes. Callers providing a `limit` argument are responsible of
+/// making subsequent calls to `validate_chain()` to complete validating the remaining
+/// blocks stored on the `block_source`.
 ///
 /// Arguments:
-/// - `parameters` Network parameters
 /// - `block_source` Source of compact blocks
-/// - `from_tip` Height & hash of last validated block; if no validation has previously
-///    been performed, this will begin scanning from `sapling_activation_height - 1`
+/// - `validate_from` Height & hash of last validated block;
+/// - `limit` Maximum number of blocks that should be validated in this call.
 ///
 /// Returns:
-/// - `Ok(())` if the combined chain is valid.
+/// - `Ok((BlockHeight, BlockHash))` if the combined chain is valid up to the given height
+/// and block hash.
 /// - `Err(Error::Chain(cause))` if the combined chain is invalid.
 /// - `Err(e)` if there was an error during validation unrelated to chain validity.
 ///
 /// This function does not mutate either of the databases.
-pub fn validate_chain<ParamsT, BlockSourceT>(
-    parameters: &ParamsT,
+pub fn validate_chain<BlockSourceT>(
     block_source: &BlockSourceT,
-    validate_from: Option<(BlockHeight, BlockHash)>,
-) -> Result<(), Error<Infallible, BlockSourceT::Error, Infallible>>
+    validate_from: (BlockHeight, BlockHash),
+    limit: Option<u32>,
+) -> Result<(BlockHeight, BlockHash), Error<Infallible, BlockSourceT::Error, Infallible>>
 where
-    ParamsT: consensus::Parameters,
     BlockSourceT: BlockSource,
 {
-    let sapling_activation_height = parameters
-        .activation_height(NetworkUpgrade::Sapling)
-        .expect("Sapling activation height must be known.");
-
     // The block source will contain blocks above the `validate_from` height.  Validate from that
     // maximum height up to the chain tip, returning the hash of the block found in the block
     // source at the `validate_from` height, which can then be used to verify chain integrity by
     // comparing against the `validate_from` hash.
-    let from_height = validate_from
-        .map(|(height, _)| height)
-        .unwrap_or(sapling_activation_height - 1);
```

Comment:
The reason that `sapling_activation_height - 1` existed in the original chain validation code that I wrote, was as a way to ensure we checked all cached CompactBlocks. We could have equivalently used 0, or any other height that was guaranteed to never be seen by the mobile SDKs (since we designed them for Sapling, and thus would never send them pre-Sapling blocks). https://github.com/zcash/librustzcash/blob/7ced7f3c5669360e9d3cfbe5622552b40406cddc/zcash_client_sqlite/src/chain.rs#L102-L130

The current code before this PR is instead interpreting it as the actual height we should start at, due in part to how `BlockSource::with_blocks` is constructed and documented. Removing the default here just moves the problem to the caller (per my above comment); we need to solve this at `BlockSource`.

Reply:
I think this is now being addressed by `blockdb_with_blocks` implementation

````Rust
let mut stmt_blocks = block_source
        .0
        .prepare(
            "SELECT height, data FROM compactblocks 
            WHERE height > ? 
            ORDER BY height ASC LIMIT ?",
        )
        .map_err(to_chain_error)?;

    let mut rows = stmt_blocks
        .query(params![
            last_scanned_height.map_or(0u32, u32::from),
            limit.unwrap_or(u32::max_value()),
        ])
        .map_err(to_chain_error)?;
````

Where the lower bound of the block source falls back to zero if `None` is provided

Reply:
Equivalent implementation can be found in `pub(crate) fn fsblockdb_with_blocks<F, DbErrT, NoteRef>` for the FsBlock block source implementation